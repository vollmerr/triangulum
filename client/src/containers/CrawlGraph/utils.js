/**
 * Stratify function
 *
 * Takes a flat array of node objects containing url and parent url fields, and processes 
 * it into a hierarchical structure, specifically an array containing a single object 
 * with a children array and so on. This mimics d3's stratify function, although is designed 
 * to create the specific formatting requried by react-d3-tree. Additionally, each node
 * object is also given a nodeSvgShape field specified by react-d3-tree for individual node
 * styling. More can be read here: https://github.com/bkrem/react-d3-tree#usage
 *
 * Params: nodes - flat array of nodes objects
 * Returns: data - array with a single object containing children objects
 */
export default function stratify(nodes) {
  
  let data = [];
  findTargetPath(nodes);
  seed(nodes[0], data);
  
  if(nodes.length > 1) {
    for(let i = 0; i < nodes.length; i++) 
      buildHierarchy(nodes[i], data[0]);
  }
  
  addStyling(data[0]);
  return data;

  // Finds target path from the nodes
  function findTargetPath(nodes) {    
    addTargetPathVar(nodes);
    if(nodes[nodes.length - 1].targetFound === 1)
      addTargetPath(nodes, nodes.length - 1, 0);

    // Assigns each object a "targetPath" key and value
    function addTargetPathVar(nodes) {
      for (let i = 0; i < nodes.length; i++)
        nodes[i].targetPath = 0;
    }
    
    // Recursivley determines target path by linear search
    function addTargetPath(nodes, chiIdx, iter) {
      nodes[chiIdx].targetPath = 1;
      if(nodes[chiIdx].parent != null) {
        for(let i = 0; i < nodes.length; i++) {
          if(nodes[chiIdx].parent === nodes[i].url)
            addTargetPath(nodes, i, iter);
        }
      }
    }
  }
  
  // Adds root node to hierarchy
  function seed(node, data) {
    addNodeObj(node, data);
  }

  // Adds new node object to end of array
  function addNodeObj(node, arr) {
    if(node.title) {
      arr.push({ 
        name: node.title,
        data: {
          url: node.url, 
          targetPath: node.targetPath,
          targetFound: node.targetFound,
        },
      });
    }
    else {
      arr.push({ 
        name: node.url,
        data: {
          url: node.url, 
          targetPath: node.targetPath,
          targetFound: node.targetFound,
        },
      });
    }
  }

  // Recursively builds hierarchy 
  function buildHierarchy(node, dataObj) {
    if(dataObj.data.url === node.parent) 
      insertChild(node, dataObj);
    else if(dataObj.children) {
      for(let i = 0; i < dataObj.children.length; i++) 
        buildHierarchy(node, dataObj.children[i]);
    }        
  }

  // Inserts child node obj onto current node
  function insertChild (node, dataObj) {
    if(!dataObj.children) 
      dataObj.children = [];
    addNodeObj(node, dataObj.children);
  }

  // Recursively goes through hierarchy adding styling object
  function addStyling(dataObj) {
    styling(dataObj);
    if(dataObj.children) {
      for(let i = 0; i < dataObj.children.length; i++) 
        addStyling(dataObj.children[i]);
    }        
  }

  // Add nodeSvgShape object to nodes for individual styling
  function styling(dataObj) {

    // Add empty objects
    dataObj.nodeSvgShape = {};
    dataObj.nodeSvgShape.shapeProps = {};

    // Set size
    if(dataObj.data.targetFound) 
      dataObj.nodeSvgShape.shapeProps.r = 20;
    else 
      dataObj.nodeSvgShape.shapeProps.r = 10;

    // Set fill
    if(dataObj.children) {
      if(dataObj.data.targetPath)
        dataObj.nodeSvgShape.shapeProps.fill = "#c0f7cc";
      else
        dataObj.nodeSvgShape.shapeProps.fill = "#e8e8e8";
    } 
    else 
      dataObj.nodeSvgShape.shapeProps.fill = "#fff";

    // Set outline color
    if(dataObj.data.targetPath) {
      if(dataObj.data.targetFound)
        dataObj.nodeSvgShape.shapeProps.stroke = "#dc3545";
      else
        dataObj.nodeSvgShape.shapeProps.stroke = "#28a745";
    }
    else
      dataObj.nodeSvgShape.shapeProps.stroke = "#b7b7b7";

    // Set outline width
    if(dataObj.data.targetFound) 
      dataObj.nodeSvgShape.shapeProps.strokeWidth = 5; 
    else 
      dataObj.nodeSvgShape.shapeProps.strokeWidth = 3;

    // Set cursor when over element to pointer
    dataObj.nodeSvgShape.shapeProps.cursor = 'pointer';
  }
}
