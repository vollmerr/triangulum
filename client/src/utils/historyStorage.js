import uuid from 'uuid/v4';

// NOTE: all the things are stored as strings in localStorage...

export const getItems = () => JSON.parse(localStorage.getItem('crawlHistory'));

export const getItem = ({ id }) => getItems()[id];

export const getCurrent = () => {
  const id = localStorage.getItem('crawlId');
  return id ? getItem({ id }) : {};
};

export const setCurrent = ({ id }) => localStorage.setItem('crawlId', id);

export const newItem = () => {
  const id = uuid();
  const crawlHistory = JSON.stringify({
    ...getItems(),
    [id]: {
      id,
      created: new Date(),
    },
  });
  localStorage.setItem('crawlHistory', crawlHistory);
  localStorage.setItem('crawlId', id);
  return id;
};

export const updateItem = ({ id, ...item }) => {
  const crawlHistory = JSON.stringify({
    ...getItems(),
    [id]: {
      ...getItem({ id }),
      ...item,
      updated: new Date(),
    },
  });
  localStorage.setItem('crawlHistory', crawlHistory);
};

export const reset = () => {
  const crawlHistory = JSON.stringify({});
  localStorage.setItem('crawlHistory', crawlHistory);
  localStorage.setItem('crawlId', '');
};

export const init = () => {
  if (!localStorage.getItem('crawlHistory')) {
    reset();
  }
};
