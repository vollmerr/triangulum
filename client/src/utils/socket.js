const host = process.env.REACT_APP_URL;

let ws = new WebSocket(host); // eslint-disable-line

ws.init = () => {
  ws = new WebSocket(host);
};

ws.onclose = ws.init;

export default ws;
