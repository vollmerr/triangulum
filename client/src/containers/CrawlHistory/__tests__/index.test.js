import React from 'react';
import { shallow } from 'enzyme';
import { Button } from 'semantic-ui-react';

import * as historyStorage from '../../../utils/historyStorage';

import { CrawlHistory } from '../index';

const props = {
  updateData: jest.fn(),
  history: {
    push: jest.fn(),
  },
};

let wrapper;
let instance;
let id;
describe('CrawlHistory', () => {
  beforeEach(() => {
    wrapper = shallow(<CrawlHistory {...props} />);
    instance = wrapper.instance();
    id = historyStorage.newItem();
  });

  describe('componentDidMount', () => {
    it('should load the history into state', () => {
      instance.loadItems = jest.fn();
      instance.componentDidMount();
      expect(instance.loadItems).toBeCalled();
    });
  });

  describe('loadItems', () => {
    it('should initalize the history and load into state', () => {
      localStorage.clear();
      instance.loadItems();
      expect(instance.state.items).toEqual({});
    });
  });

  describe('onReset', () => {
    it('should reset the history', () => {
      instance.loadItems();
      instance.onReset();
      expect(instance.state.items).toEqual({});
    });

    it('should force a re-render', () => {
      instance.forceUpdate = jest.fn();
      instance.onReset();
      expect(instance.forceUpdate).toHaveBeenCalled();
    });
  });

  describe('onClick', () => {
    it('should update the graphs data', () => {
      instance.onClick(null, { id });
      const data = historyStorage.getItem({ id });
      expect(props.updateData).toBeCalledWith({ data });
    });

    it('should redirect to the graph page', () => {
      instance.onClick(null, { id });
      expect(props.history.push).toBeCalledWith('/graph');
    });
  });

  describe('render', () => {
    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should reset the history when the reset button is pressed', () => {
      historyStorage.newItem();
      instance.loadItems();
      wrapper.find(Button).simulate('click');
      const items = historyStorage.getItems();
      expect(items).toEqual({});
    });
  });
});
