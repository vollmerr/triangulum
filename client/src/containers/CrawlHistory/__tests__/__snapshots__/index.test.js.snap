// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CrawlHistory render should render correctly 1`] = `
<Container
  className="page"
  text={true}
>
  <HistoryItems
    items={Object {}}
    onClick={[Function]}
  />
</Container>
`;
